chain/rust/src/genesis/shelley/parse.rs

```diff
@@ -0,0 +1,227 @@
+
+use super::{config, raw::{self}};
+
+#[derive(Debug, thiserror::Error)]
+pub enum GenesisJSONError {
+    #[error("JSON: {0:?}")]
+    Serde(#[from] serde_json::Error),
+    #[error("Crypto: {0:?}")]
+    CryptoError(#[from] CryptoError),
+    #[error("ParseInt: {0:?}")]
+    ParseInt(#[from] std::num::ParseIntError),
+    #[error("ParseIP: {0:?}")]
+    ParseIP(#[from] crate::certs::utils::IPStringParsingError),
+    #[error("Unexpected network type: {0:?}")]
+    ParseNetwork(String),
+}
+
+pub fn parse_genesis_data<R: Read>(json: R) -> Result<config::ShelleyGenesisData, GenesisJSONError> {
+    let data_value: serde_json::Value = serde_json::from_reader(json)?;
+    let data: raw::ShelleyGenesisData = serde_json::from_value(data_value.clone())?;
+
+    let mut initial_funds = BTreeMap::new();
+    for (addr_hex, balance) in &data.initialFunds {
+        initial_funds.insert(
+            Address::from_hex(addr_hex)
+                .map_err(CryptoError::from)?,
+            *balance,
+        );
+    }
+
+    let network_id = match data.networkId.as_str() {
+        "Mainnet" => crate::NetworkId::mainnet().network,
+        "Testnet" => crate::NetworkId::testnet().network,
+        val => return Err(GenesisJSONError::ParseNetwork(val.to_string())),
+    };
+
+    let staking = match data.staking.as_ref() {
+        Some(raw) => {
+            // 1) Get stake pools
+            let mut pools: BTreeMap<Ed25519KeyHash, PoolParams> = BTreeMap::new();
+            for (pool_id, params) in &raw.pools {
+                let ration = fraction::Fraction::from_str(&params.margin).unwrap();
+                let mut owners = Vec::<Ed25519KeyHash>::new();
+                for owner in & params.owners {
+                    owners.push(Ed25519KeyHash::from_hex(owner)?);
+                }
+                let mut relays = Vec::<Relay>::new();
+                for relay in & params.relays {
+                    if let Some((key, value)) = relay.iter().next() {
+                        match key.as_str() {
+                            "single host address" => {
+                                let ipv4 = match value.IPv4.as_ref() {
+                                    Some(s) => Some(Ipv4::from_str(s)?),
+                                    _ => None
+                                };
+                                let ipv6 = match value.IPv6.as_ref() {
+                                    Some(s) => Some(Ipv6::from_str(s)?),
+                                    _ => None
+                                };
+                                relays.push(Relay::new_single_host_addr(
+                                    value.port,
+                                    ipv4,
+                                    ipv6
+                                ));
+                            },
+                            _ => panic!("Only single host address relays are supported in cardano-node Relay JSON parsing")
```

Comment:
If we're going to panic here it'd be good to say so in the function comment.

Edit: I read the comment on `RelayTypeMap` saying you couldn't even find JSON for other types so maybe this is okay.